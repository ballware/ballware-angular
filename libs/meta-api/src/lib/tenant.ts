import { HttpClient } from '@angular/common/http';
import { map, Observable } from 'rxjs';

import { parse } from 'json5/lib';

import { CompiledTenant, NavigationLayout } from '@ballware/meta-model';

/**
 * Interface for tenant data operations
 */
 export interface MetaTenantApi {
  /**
   * Fetch metadatan for tenant
   * @param tenant Identifier of tenant
   * @returns Observable containing compiled tenant metadata
   */
  metadataForTenant: (tenant: string) => Observable<CompiledTenant>;
}

interface Tenant {
  Id: string;
  Name: string;
  Navigation?: string;
  RightsCheckScript?: string;
}

const compileTenant = (tenant: Tenant): CompiledTenant => {
  const compiledTenant = {
    id: tenant.Id,
    name: tenant.Name,
    navigation: tenant.Navigation
      ? (parse(tenant.Navigation) as NavigationLayout)
      : ({} as NavigationLayout),
  } as CompiledTenant;

  if (tenant.RightsCheckScript) {
    const compiledArgs = ['userinfo', 'right'];
    const compiledFn = Function.apply(
      Function,
      compiledArgs.concat(tenant.RightsCheckScript)
    );

    compiledTenant.hasRight = compiledFn
      ? (userinfo, right) => compiledFn.apply(compiledFn, [userinfo, right])
      : () => true;
  }

  return compiledTenant;
};

const metadataFunc = (http: HttpClient, serviceBaseUrl: string) => (
  tenant: string
): Observable<CompiledTenant> => {
  const url = `${serviceBaseUrl}/api/tenant/metadatafortenant/${tenant}`;

  return http
    .get<Tenant>(url)
    .pipe(map(data => compileTenant(data)));
};

/**
 * Create adapter for tenant fetch operations with ballware.meta.service
 * @param serviceBaseUrl Base URL to connect to ballware.meta.service
 * @returns Adapter object providing data operations
 */
export function createMetaBackendTenantApi(
  httpClient: HttpClient, 
  serviceBaseUrl: string
): MetaTenantApi {
  return {
    metadataForTenant: metadataFunc(httpClient, serviceBaseUrl),
  } as MetaTenantApi;
}
