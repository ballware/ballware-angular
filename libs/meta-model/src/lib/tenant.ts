import { Template } from "./template";

/**
 * Navigation tree item
 */
 export interface NavigationLayoutItem {
  /**
   * Type of navigation item
   */
  type: 'page' | 'group' | 'section';

  /**
   * Options
   */
  options: {
    /**
     * Page identifier if item type is 'page'
     */
    page?: string;

    /**
     * Optional url for external page not part of application
     */
    url?: string;

    /**
     * Caption for navigation item type 'page' and 'group'
     */
    caption?: string;

    /**
     * Optional icon for navigation item type 'page' and 'group'
     */
     icon?: string;
  };

  /**
   * Collection of child items (only for 'group' and 'section')
   */
  items?: Array<NavigationLayoutItem>;
}

/**
 * Navigation layout for single tenant
 */
export interface NavigationLayout {
  /**
   * Headline for tenant application
   */
  title: string;

  /**
   * Default page activated on tenant entered
   */
  defaultUrl?: string;

  /**
   * Collection of layout items
   */
  items: Array<NavigationLayoutItem>;
}

/**
 * Metadata for tenant
 */
export interface CompiledTenant {
  /**
   * Unique identifier of tenant
   */
  id: string;

  /**
   * Display name of tenant
   */
  name: string;

  /**
   * Navigation metadata for tenant
   */
  navigation?: NavigationLayout;

  /**
   * List of tenant templates
   */
  templates?: Array<Template>;

  /**
   * Custom script for access rights check of tenant
   * @param userinfo Assigned user rights
   * @param right Requested right
   * @returns true if access allowed, false if access denied
   */
  hasRight: (userinfo: Record<string, unknown>, right: string) => boolean;
}
