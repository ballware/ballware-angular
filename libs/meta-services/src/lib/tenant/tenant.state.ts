import { CompiledTenant, NavigationLayout, NavigationLayoutItem } from "@ballware/meta-model";
import { createFeatureSelector, createSelector } from "@ngrx/store";

export const tenantFeatureKey = "tenant";

export interface TenantState {
    tenant?: CompiledTenant,
    title?: string,
    navigationLayout?: NavigationLayout,
    navigationTree?: Record<string, unknown>[],
    pages?: NavigationLayoutItem[],
    hasRight?: (rights: string) => boolean
}

const selectTenantFeature = createFeatureSelector<TenantState>(tenantFeatureKey);

export const selectTenant = createSelector(selectTenantFeature, (state: TenantState) => state.tenant);
export const selectTitle = createSelector(selectTenantFeature, (state: TenantState) => state.title);
export const selectNavigationLayout = createSelector(selectTenantFeature, (state: TenantState) => state.navigationLayout);
export const selectNavigationTree = createSelector(selectTenantFeature, (state: TenantState) => state.navigationTree);
export const selectPages = createSelector(selectTenantFeature, (state: TenantState) => state.pages);
export const selectHasRight = createSelector(selectTenantFeature, (state: TenantState) => state.hasRight);
