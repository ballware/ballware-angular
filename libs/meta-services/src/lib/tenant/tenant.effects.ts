import { inject } from "@angular/core";
import { Actions, createEffect, ofType } from "@ngrx/effects";
import { MetaApiService } from "@ballware/meta-api";
import { identityUserLogin } from "../identity/identity.actions";
import { combineLatest, map, of, switchMap } from "rxjs";
import { tenantFetched } from "./tenant.actions";

export const fetchTenant = createEffect((actions$ = inject(Actions), metaApiService = inject(MetaApiService)) => 
    actions$.pipe(ofType(identityUserLogin))                
        .pipe(switchMap((user) => combineLatest([of(user), metaApiService.metaTenantApi.metadataForTenant(user.tenant)]) ))
        .pipe(map(([user, compiledTenant]) => tenantFetched({ user: user.currentUser, tenant: compiledTenant })))
, { functional: true, dispatch: true });