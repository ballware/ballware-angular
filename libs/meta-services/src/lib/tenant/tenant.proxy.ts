import { Store } from "@ngrx/store";
import { TenantService } from "../tenant.service";
import { selectHasRight, selectNavigationLayout, selectNavigationTree, selectPages, selectTemplates, selectTenant, selectTitle } from "./tenant.state";

export class TenantServiceProxy implements TenantService {
    constructor(private store: Store) {}

    public get tenant$() {
        return this.store.select(selectTenant);
    }

    public get title$() {
        return this.store.select(selectTitle);
    }

    public get navigationLayout$() {
        return this.store.select(selectNavigationLayout);
    }

    public get navigationTree$() {
        return this.store.select(selectNavigationTree);
    }

    public get pages$() {
        return this.store.select(selectPages);
    }

    public get tenantTemplates$() {
        return this.store.select(selectTemplates);
    }

    public get hasRight$() {
        return this.store.select(selectHasRight);
    }
}