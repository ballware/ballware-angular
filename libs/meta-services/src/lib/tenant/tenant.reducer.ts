import { NavigationLayout, NavigationLayoutItem } from "@ballware/meta-model";
import { createReducer, on } from "@ngrx/store";
import { NavigationTreeItem } from "../tenant.service";
import { tenantFetched } from "./tenant.actions";
import { TenantState } from "./tenant.state";

export const initialState = {
    
} as TenantState;


export const buildNavigationTree = (hasRight: (right: string) => boolean, navigation: NavigationLayout): NavigationTreeItem[] => {

    const collectAllowedItems = (items: NavigationLayoutItem[]): NavigationTreeItem[] => {
      const mappedItems = [] as NavigationTreeItem[];
  
      items?.forEach(item => {
        if (item.type === 'page' && item.options?.page) {
          const pageVisible = hasRight(`generic.page.${item.options.page}`);
  
          if (pageVisible) {
            mappedItems.push({
              type: 'page',
              text: item.options?.caption,
              icon: item.options?.icon,
              url: `/page/${item.options.url}`
            });
          }
        } else if (item.type === 'section' && item.items) {
          const subItems = collectAllowedItems(item.items);
            
          if (subItems.length > 0) {
            /*
            if (mappedItems.length > 0) {
              mappedItems.push({
                type: 'section',
                html: '<hr/>',
                disabled: true
              });
            }
            */
  
            mappedItems.push(...subItems);
          }          
        } else if (item.type === 'group' && item.items) {
          const subItems = collectAllowedItems(item.items);
  
          if (subItems.length > 0) {
            mappedItems.push({
              expanded: false,
              type: 'group',
              text: item.options?.caption,
              icon: item.options?.icon,
              items: subItems
            });
          }
        }
      });
  
      return mappedItems;
    }
  
    return collectAllowedItems(navigation?.items ?? []);
  }
  
  export const buildPageList = (hasRight: (right: string) => boolean, navigation: NavigationLayout): NavigationLayoutItem[] => {
  
    const collectPages = (items: NavigationLayoutItem[]): NavigationLayoutItem[] => {
      const pages = [] as NavigationLayoutItem[];
  
      items?.forEach(item => {
        if (item.type === 'page' && item.options?.page) {
          const pageVisible = hasRight(`generic.page.${item.options.page}`);
  
          if (pageVisible) {
            pages.push(item);
          }
        } else if ((item.type === 'section' || item.type === 'group') && item.items) {
          pages.push(...collectPages(item.items));
        }
      });
  
      return pages;
    }
  
    return collectPages(navigation?.items ?? []);
  }
  

export const tenantReducer = createReducer(
    initialState,     
    on(tenantFetched, (state, { user, tenant }) => ({
        ...state,
        tenant,
        title: tenant?.name,
        hasRight: tenant?.hasRight ? (right) => tenant.hasRight(user, right) : undefined,
        navigationLayout: tenant?.navigation,
        navigationTree: tenant?.navigation ? buildNavigationTree((right) => tenant?.hasRight(user, right), tenant?.navigation) : undefined,
        pages: tenant?.navigation ? buildPageList((right) => tenant?.hasRight(user, right), tenant.navigation) : undefined,
        templates: tenant?.templates ?? []
    }))    
);