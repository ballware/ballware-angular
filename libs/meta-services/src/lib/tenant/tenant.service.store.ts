import { Injectable } from "@angular/core";
import { TenantService } from "../tenant.service";
import { Store } from "@ngrx/store";
import { selectHasRight, selectNavigationLayout, selectNavigationTree, selectPages, selectTenant, selectTitle } from "./tenant.state";

@Injectable()
export class TenantServiceStore extends TenantService {
    constructor(private store: Store) {
        super();
    }

    public get tenant$() {
        return this.store.select(selectTenant);
    }

    public get title$() {
        return this.store.select(selectTitle);
    }

    public get navigationLayout$() {
        return this.store.select(selectNavigationLayout);
    }

    public get navigationTree$() {
        return this.store.select(selectNavigationTree);
    }

    public get pages$() {
        return this.store.select(selectPages);
    }

    public get hasRight$() {
        return this.store.select(selectHasRight);
    }

        /*
                  this.store.dispatch(identityRefreshToken());
            }
          },
          {
            text: this.translationService.transform('session.manageaccount'),
            onClick: () => {
              this.accountMenu?.instance.hide();
              this.store.dispatch(identityManageProfile());
            }
          },
          {
            text: this.translationService.transform('session.logout', { user: userName }),
            onClick: () => {
              this.accountMenu?.instance.hide();
              this.store.dispatch(identityUserLogout());
              */
}