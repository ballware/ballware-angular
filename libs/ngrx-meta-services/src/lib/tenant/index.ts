import { StoreModule } from '@ngrx/store';
import { tenantReducer } from './tenant.reducer';
import { tenantFeatureKey } from './tenant.state';
import { EffectsModule } from '@ngrx/effects';
import { fetchTenant } from './tenant.effects';

export * from './tenant.actions';
export * from './tenant.state';

export const TenantFeatureModule = StoreModule.forFeature(tenantFeatureKey, tenantReducer);
export const TenantEffectsModule = EffectsModule.forFeature({ fetchTenant });