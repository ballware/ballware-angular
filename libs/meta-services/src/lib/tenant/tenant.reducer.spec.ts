import { createAction } from "@ngrx/store";
import { buildNavigationTree, buildPageList, initialState, tenantReducer } from "./tenant.reducer";
import { CompiledTenant, NavigationLayout } from "@ballware/meta-model";
import { tenantFetched } from "./tenant.actions";

const navigationLayout = {
    items: [
        { 
            type: 'group',             
            options: {
                caption: 'allowed_group'
            },       
            items: [
                {
                    type: 'page',
                    options: {
                        page: 'allowed_page_1',
                        url: 'page_1',
                        caption: 'Page 1'
                    }
                },
                {
                    type: 'page',
                    options: {
                        url: 'page_unknown',
                        caption: 'Unknown page'
                    }
                },
                {
                    type: 'section',
                    items: [
                        {
                            type: 'page',
                            options: {
                                page: 'allowed_page_2',
                                url: 'page_2',
                                caption: 'Page 2'
                            }
                        },
                        {
                            type: 'page',
                            options: {
                                page: 'forbidden_page_3',
                                url: 'page_3',
                                caption: 'Page 3'
                            }
                        }
                    ]
                },                        
            ]
        },
        { 
            type: 'group',             
            options: {
                caption: 'forbidden_group'
            },       
            items: [
                {
                    type: 'page',
                    options: {
                        page: 'forbidden_page_4',
                        url: 'page_4',                                
                        caption: 'Page 4'
                    }
                },
                {
                    type: 'section',
                    items: [
                        {
                            type: 'page',
                            options: {
                                page: 'forbidden_page_5',
                                url: 'page_5',
                                caption: 'Page 5'
                            }
                        },
                        {
                            type: 'page',
                            options: {
                                page: 'forbidden_page_6',
                                url: 'page_6',
                                caption: 'Page 6'
                            }
                        }
                    ]
                }
            ]
        }
    ]
} as NavigationLayout;


describe('build navigation tree', () => {
    it('empty metadata should return empty navigation tree', () => {
        const emptyNavigationLayout = {} as NavigationLayout;

        const navigationTree = buildNavigationTree(() => true, emptyNavigationLayout);

        expect(navigationTree).toEqual([]);
    });

    it('metadata with not allowed pages should generate tree without excluded pages and empty groups', () => {

        const navigationTree = buildNavigationTree((right) => !right.includes('forbidden'), navigationLayout);

        expect(navigationTree).toMatchSnapshot();
    });
});

describe('build page list', () => {
    it('empty metadata should return empty page list', () => {
        const emptyNavigationLayout = {} as NavigationLayout;

        const pageList = buildPageList(() => true, emptyNavigationLayout);

        expect(pageList).toEqual([]);
    });

    it('metadata with not allowed pages should generate page list without excluded pages', () => {

        const pageList = buildPageList((right) => !right.includes('forbidden'), navigationLayout);

        expect(pageList).toMatchSnapshot();
    });
});

describe('reducer', () => {
    it('reducer should return init state', () => {
        const noopAction = createAction('noop');

        const newState = tenantReducer(undefined, noopAction);
        
        expect(newState).toEqual(initialState);
    });

    it('fetch tenant without navigation should add fetched tenant to state', () => {
        const user = { 'preferred_username': 'John Doo' } as Record<string, unknown>;
        const tenant = { 
            name: 'My tenant'            
        } as CompiledTenant;

        const newState = tenantReducer(initialState, tenantFetched({ user, tenant }));

        expect(newState.title).toEqual(tenant.name);
        expect(newState.tenant).toEqual(tenant);
        expect(newState.navigationLayout).toBeUndefined();
        expect(newState.navigationTree).toBeUndefined();
        expect(newState.pages).toBeUndefined();
        expect(newState.hasRight).toBeUndefined();
    });

    it('fetch tenant should add fetched tenant to state', () => {
        const user = { 'preferred_username': 'John Doo' } as Record<string, unknown>;
        const tenant = { 
            name: 'My tenant',
            hasRight: (_user, _right) => true,
            navigation: navigationLayout
        } as CompiledTenant;

        const newState = tenantReducer(initialState, tenantFetched({ user, tenant }));

        expect(newState.title).toEqual(tenant.name);
        expect(newState.tenant).toEqual(tenant);
        expect(newState.hasRight).toBeDefined();
        expect(newState.hasRight && newState.hasRight('nonsense')).toEqual(true);
    });
});