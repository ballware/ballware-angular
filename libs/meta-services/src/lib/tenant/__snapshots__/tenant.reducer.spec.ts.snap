// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build navigation tree metadata with not allowed pages should generate tree without excluded pages and empty groups 1`] = `
[
  {
    "expanded": false,
    "icon": undefined,
    "items": [
      {
        "icon": undefined,
        "text": "Page 1",
        "type": "page",
        "url": "/page/page_1",
      },
      {
        "icon": undefined,
        "text": "Page 2",
        "type": "page",
        "url": "/page/page_2",
      },
    ],
    "text": "allowed_group",
    "type": "group",
  },
]
`;

exports[`build page list metadata with not allowed pages should generate page list without excluded pages 1`] = `
[
  {
    "options": {
      "caption": "Page 1",
      "page": "allowed_page_1",
      "url": "page_1",
    },
    "type": "page",
  },
  {
    "options": {
      "caption": "Page 2",
      "page": "allowed_page_2",
      "url": "page_2",
    },
    "type": "page",
  },
]
`;
